c2rust-analyze/src/util.rs

```diff
@@ -2,11 +2,43 @@ use crate::labeled_ty::LabeledTy;
 use rustc_hir::def::DefKind;
 use rustc_hir::def_id::DefId;
 use rustc_middle::mir::{
-    Field, Local, Mutability, Operand, PlaceElem, PlaceRef, ProjectionElem, Rvalue,
+    Field, Local, Mutability, Operand, Place, PlaceElem, PlaceRef, ProjectionElem, Rvalue,
 };
 use rustc_middle::ty::{AdtDef, DefIdTree, SubstsRef, Ty, TyCtxt, TyKind, UintTy};
+use std::collections::HashMap;
 use std::fmt::Debug;
 
+/// A mapping for substituting [`Place`]s adhering to the
+/// following pattern
+///
+/// ```mir
+/// _1 = malloc(...);
+/// _2 = _1 as *mut T;
+/// ```
+///
+/// where `_1` is [`*c_void`](core::ffi::c_void).
+///
+/// In this case, `_1` would be mapped to `_2`, which is indicative
+/// of the amended statement:
+///
+/// ```mir
+/// _2 = malloc(...);
+/// ```
+#[derive(Default)]
+pub struct CVoidCasts<'tcx>(pub HashMap<Place<'tcx>, Place<'tcx>>);
```

Comment:
Should the two mappings, the alloc one and the free one, be conflated?  Or should there be two maps?

Reply:
For now it makes no difference. We treat casts to/from `*c_void` the same. When we start to treat them differently then it may make sense to separate them.

Reply:
We don't really treat them totally the same.

In

```rust
    pub fn contains(&self, lhs: &Place<'tcx>, rv: &Rvalue<'tcx>) -> bool {
        matches!(rv, Rvalue::Cast(_, Operand::Copy(p) | Operand::Move(p), _) if self.0.contains_key(p))
            || self.0.contains_key(lhs)
    }
```

one `.contains_key` is for to casts and one is for from casts.  Couldn't this potentially lead to a false positive?

Reply:
I agree with @kkysen that conflating casts from `void*` and casts to `void*` seems risky.  Here's an example:
```Rust
x = malloc(...);
p = x as *mut u8;
p = y;
free(x);
```
Here, `x` is cast to `p` after the `malloc`, so the map will contain an entry `(x, p)`.  There is no cast whose LHS is `x`, so the `free` won't add any entries to the map.  But when processing the `free(x)`, we'll see the entry `(x, p)` and treat it as `free(p)`, which is incorrect.

To reiterate from my comment above, I think the maps should be keyed on the MIR `Location` of the `malloc`/`free` so there can't be any ambiguity arising from the same `Place` appearing in more than one call.

Reply:
Would you say #826 covered this adequately?